src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -195,7 +202,7 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
         private void CaptureVariable(Symbol variable, SyntaxNode syntax)
         {
             var type = (variable.Kind == SymbolKind.Local) ? ((LocalSymbol)variable).Type : ((ParameterSymbol)variable).Type;
-            if (type.IsRestrictedType())
+            if (type.IsRestrictedType() || variable is LocalSymbol { RefKind: not RefKind.None, IsCompilerGenerated: false })
```

Comment:
>IsCompilerGenerated: false [](http://example.com/codeflow?start=96&length=26)

Why are we letting compiler generated locals through here?

Reply:
Existing codegen expects some spilled ref locals to be in the list of hoisted variables. Although it's not all compiler generated locals, it's just these ones:

[https://github.com/dotnet/roslyn/blob/92fb904fb7200ce46c3b07bfe3835a3c955613ca/src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs#L172-L173](https://github.com/dotnet/roslyn/blob/92fb904fb7200ce46c3b07bfe3835a3c955613ca/src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.cs#L172-L173)

I will make the check here narrower.